src/rt/minfo.d

```diff
@@ -62,174 +62,382 @@ struct ModuleGroup
+        // check for cycles. If a cycle is found, the cycle path is transformed
+        // into a string and thrown as an error.
+        //
+        // Each call into this function is given a module that has static
+        // ctor/dtors that must be dealt with. It recurses only when it finds
+        // dependencies that also have static ctor/dtors.
+        void processMod(int curidx)
+        {
+            assert(curidx != -1);
+            immutable ModuleInfo* current = _modules[curidx];
+
+            // First, determine what modules are reachable.
+            auto reachable = cast(size_t *)alloca(flagbytes);
+            auto cycleIdx = findDeps(curidx, reachable);
+            if(cycleIdx != -1)
+            {
+                auto cycleMod = _modules[cycleIdx];
+                // found a cycle
+                println("Cyclic dependency between module ", cycleMod.name, " and ", current.name);
+                genPath(cycleIdx, curidx);
+
+                foreach(midx; cyclePath[0 .. $-1])
+                {
+                    println(_modules[midx].name, bt(relevant, midx) ? "* ->" : " ->");
+                }
+                println(cycleMod.name, "*");
+                throw new Error(errmsg);
+            }
+
+            // process the dependencies. First, we process all relevant ones
+            bts(ctorstart, curidx);
+            foreach (int i; 0 .. len)
+            {
+                if(i != curidx && bt(reachable, i) &&
+                   bt(relevant, i) && !bt(ctordone, i))
+                {
+                    assert(!bt(ctorstart, i)); // sanity check, this should have been flagged a cycle earlier
+                    processMod(i);
+                }
+            }
+
+            // now mark this node, and all nodes reachable from this module as done.
+            bts(ctordone, curidx);
+            btr(ctorstart, curidx);
+            foreach (int i; 0 .. len)
+            {
+                if(bt(reachable, i))
+                {
+                    // Since relevant dependencies are already marked as done
+                    // from recursion above, no reason to check for relevance,
+                    // that is a wasted op.
+                    bts(ctordone, i);
+                }
+            }
+
+            // add this module to the construction order list
+            ctors[ctoridx++] = current;
+        }
+
+        immutable(ModuleInfo)*[]
+            doSort(int function(immutable ModuleInfo *) getf)
+        {
+            clearFlags(relevant);
+            clearFlags(ctorstart);
+            clearFlags(ctordone);
+
+            // pre-allocate enough space to hold all modules.
+            ctors = (cast(immutable(ModuleInfo)**).malloc(len * (void*).sizeof))[0 .. len];
+            ctoridx = 0;
+            foreach (int idx, m; _modules)
+            {
+                // TODO: Should null ModuleInfo be allowed?
+                if (m is null) continue;
```

Comment:
Well, looking back at the current released code, null modules weren't checked for, and would definitely have crashed the runtime. Are you sure?